libraries/AP_HAL_ChibiOS/SPIDevice.cpp

```diff
@@ -0,0 +1,411 @@
+            break;
+        case 1:
+            spi_clock_freq = SPI2_CLOCK;
+            break;
+        case 2:
+            spi_clock_freq = SPI4_CLOCK;
+            break;
+        case 3:
+            spi_clock_freq = SPI4_CLOCK;
+            break;
+        default:
+            spi_clock_freq = SPI1_CLOCK;
+            break;
+    }
+
+    for(i = 0; i <= 7; i++) {
+        spi_clock_freq /= 2;
+        if (spi_clock_freq <= _frequency) {
+            break;
+        }
+    }
+    switch(i) {
+        case 0: //PCLK DIV 2
+            return 0;
+        case 1: //PCLK DIV 4
+            return SPI_CR1_BR_0;
+        case 2: //PCLK DIV 8
+            return SPI_CR1_BR_1;
+        case 3: //PCLK DIV 16
+            return SPI_CR1_BR_1 | SPI_CR1_BR_0;
+        case 4: //PCLK DIV 32
+            return SPI_CR1_BR_2;
+        case 5: //PCLK DIV 64
+            return SPI_CR1_BR_2 | SPI_CR1_BR_0;
+        case 6: //PCLK DIV 128
+            return SPI_CR1_BR_2 | SPI_CR1_BR_1;
+        case 7: //PCLK DIV 256
+        default:
+            return SPI_CR1_BR_2 | SPI_CR1_BR_1 | SPI_CR1_BR_0;
+    }
+}
+
+bool SPIDevice::transfer(const uint8_t *send, uint32_t send_len,
+                         uint8_t *recv, uint32_t recv_len)
+{
+    if (send_len == recv_len && send == recv) {
+        // simplest cases, needed for DMA
+        do_transfer(send, recv, recv_len);
+        return true;
+    }
+    uint8_t buf[send_len+recv_len];
+    if (send_len > 0) {
+        memcpy(buf, send, send_len);
+    }
+    if (recv_len > 0) {
+        memset(&buf[send_len], 0, recv_len);
+    }
+    do_transfer(buf, buf, send_len+recv_len);
+    if (recv_len > 0) {
+        memcpy(recv, &buf[send_len], recv_len);
+    }
+    return true;
+}
+
+bool SPIDevice::transfer_fullduplex(const uint8_t *send, uint8_t *recv, uint32_t len)
+{
+    uint8_t buf[len];
+    memcpy(buf, send, len);
+    do_transfer(buf, buf, len);
+    memcpy(recv, buf, len);
+    return true;
+}
+
+AP_HAL::Semaphore *SPIDevice::get_semaphore()
+{
+    return &bus.semaphore;
+}
+
+
+AP_HAL::Device::PeriodicHandle SPIDevice::register_periodic_callback(uint32_t period_usec, AP_HAL::Device::PeriodicCb cb)
+{
+    return bus.register_periodic_callback(period_usec, cb, this);
+}
+
+bool SPIDevice::adjust_periodic_callback(AP_HAL::Device::PeriodicHandle h, uint32_t period_usec)
+{
+    return bus.adjust_timer(h, period_usec);
+}
+
+/*
+  allow for control of SPI chip select pin
+ */
+bool SPIDevice::set_chip_select(bool set)
+{
+    if (set && cs_forced) {
+        return true;
+    }
+    if (!set && !cs_forced) {
+        return false;
+    }
+    if (!set && cs_forced) {
+        spiUnselectI(spi_devices[device_desc.bus].driver);          /* Slave Select de-assertion.       */
+        spiReleaseBus(spi_devices[device_desc.bus].driver);              /* Ownership release.               */
+        cs_forced = false;
+        bus.dma_handle->unlock();
+    } else {
+        bus.dma_handle->lock();
```

Comment:
there is lock here but no unlock after

Reply:
its correct, the caller is handling the lock/unlock